test/base/Escrow/Dapps/UniswapSwapTest.js

```diff
@@ -1,59 +1,158 @@
 
 contract("UniswapSwapTest", function(accounts) {
-  const uniswapEncorder = new UniswapEncoder(web3);
+  const DONT_ALTER_BALANCE = 999999;
+  const SIMULATE_UNISWAP_RESPONSE_ERROR = 777777;
 
   let uniswapV2Router02;
-  let uniswap;
+  let instance;
   let dai;
   let usdc;
 
-  before(async () => {
-    uniswapV2Router02 = await Mock.new();
-    await uniswapV2Router02.givenMethodReturn(
-      uniswapEncorder.encodeRouterTokensForTokens(),
-      abi.rawEncode([ "uint256" ], [ 1 ])
-    );
-    await uniswapV2Router02.givenMethodReturn(
-      uniswapEncorder.encodeRouterTokensForETH(),
-      abi.rawEncode([ "uint256" ], [ 1 ])
-    );
-    await uniswapV2Router02.givenMethodReturn(
-      uniswapEncorder.encodeRouterETHForTokens(),
-      abi.rawEncode([ "uint256" ], [ 1 ])
-    );
-
-    uniswap = await Uniswap.new(uniswapV2Router02.address);
-
-    dai = await DAI.new();
-    usdc = await USDC.new();
+  beforeEach(async () => {
+    dai = await DAI.new(); // Used as Input as well as ETH
+    usdc = await USDC.new(); // Used as Input as well as ETH
+    weth = await WETH.new(); // Used only as Output
+    uniswapV2Router02 = await UniswapRouter.new();
+    instance = await Uniswap.new();
   });
 
-  beforeEach(async () => {
+  withData({
+    _1_ethForTokens: [ 0, [ 'eth', 'dai' ], 50, 50, 4, true, true, false, null ],    
+    _2_wethIsNotContract: [ 0, [ 'eth' ], 0, 0, 0, true, false, true, "CANONICAL_WETH_MUST_BE_CONTRACT" ],
+    _3_routerIsNotContract: [ 0, [ 'eth' ], 0, 0, 0, false, true, true, "ROUTER_MUST_BE_A_CONTRACT" ],
+    _4_pathTooShort: [ 0, [ 'eth' ], 0, 0, 0, true, true, true, "UNISWAP_PATH_TOO_SHORT" ],
+    _5_sourceAndDestinationSame: [ 0, [ 'dai', 'dai' ], 0, 0, 0, true, true, true, "UNISWAP_SOURCE_AND_DESTINATION_SAME" ],
+    _6_minDestinationZero: [ 0, [ 'eth', 'dai' ], 0, 0, 0, true, true, true, "UNISWAP_MIN_DESTINATION_ZERO" ],
+    _7_insufficientSourceEth: [ 0, [ 'eth', 'usdc' ], 50, 0, 1, true, true, true, "UNISWAP_INSUFFICIENT_ETH" ],
+    _8_insufficientSourceToken: [ 0, [ 'dai', 'usdc' ], 50, 0, 1, true, true, true, "UNISWAP_INSUFFICIENT_TOKENS" ],
+    _9_ethForTokensUniswapError: [ 0, [ 'eth', 'dai' ], 50, 50, SIMULATE_UNISWAP_RESPONSE_ERROR, true, true, true, "UNISWAP_ERROR_SWAPPING" ],
+    _10_ethForTokensBalanceError: [ 0, [ 'eth', 'dai' ], 50, 50, DONT_ALTER_BALANCE, true, true, true, "UNISWAP_BALANCE_NOT_INCREASED" ],
+   }, function(
+    senderAccount,
+    path,
+    sourceAmount,
+    sourceBalance,
+    minDestination,
+    routerIsContract,
+    wethIsContract,
+    mustFail,
+    expectedErrorMessage
+  ) {
+    it(t("uniswap", "swapExactETHForTokens", "Should be able (or not) to swap tokens on Uniswap", mustFail), async function() {
+      // Setup
+      if (sourceBalance > 0) {
+        if (path[0] === 'eth') {
+          await web3.eth.sendTransaction({ to: instance.address, from: accounts[0], value: sourceBalance + 1 })
```

Comment:
As mentioned above, we don't the dapp contracts to accept any funds. For this test it makes sense, however, instead of having a fallback function on the dapp contract itself, we should create a mock contract that extends it with just a payable fallback function to accept ETH so we don't deploy the main dapp with one.

Reply:
Same as above, payable fallback moved to Mock only.